File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03299/p03299_num2_parsed.py
import sys

MOD = 10**9 + 7

def solve(N, h):
    dp = [[0] * 4 for _ in range(N)]
    dp[0][0] = 1
    dp[0][1] = 1
    dp[0][2] = 1
    dp[0][3] = 1

    for i in range(1, N):
        dp[i][0] = (dp[i-1][0] + dp[i-1][2] + dp[i-1][3]) % MOD
        dp[i][1] = (dp[i-1][1] + dp[i-1][2] + dp[i-1][3]) % MOD
        dp[i][2] = (dp[i-1][0] + dp[i-1][1]) % MOD
        dp[i][3] = (dp[i-1][0] + dp[i-1][1]) % MOD

    ans = 1
    for i in range(N):
        if h[i] != h[i-1]:
            ans = (ans * dp[i][3]) % MOD
        else:
            ans = (ans * dp[i][2]) % MOD

    return ans

def main():
    input=sys.stdin.readline
    N = int(input())
    h = list(map(int, input().split()))
    ans = solve(N, h)
    print(ans)

if __name__ == "__main__":
    main()

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03299/p03299_num2_parsed.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1000000007;

int main() {
    int N;
    cin >> N;

    vector<int> h(N);
    for (int i = 0; i < N; i++) {
        cin >> h[i];
    }

    long long ans = 1;
    for (int i = 0; i < N; i++) {
        if (i == 0) {
            ans *= h[i] + 1;
        } else if (i == N - 1) {
            ans *= h[i - 1] + 1;
        } else {
            ans *= min(h[i - 1], h[i]) + 1;
        }
        ans %= MOD;
    }

    cout << ans << endl;

    return 0;
}